design/0001-metallb-crd.md

```diff
@@ -0,0 +1,173 @@
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Testing](#testing)
+- [Debugging and Troubleshooting](#debugging-and-troubleshooting)
+<!-- /toc -->
+
+## Summary
+Today MetalLB allows only configuration of layer2 and BGP protocols using Configmap objects, 
+which is simple from configuration perspective and doesn't require cluster-admin rights,
+however it doesn't have validation on the API server level, no support for API versioning and 
+no fine-grained permission model via RBAC.
+
+## Motivation
+Custom resource definition "CRD" takes advantage of kubernetes extensibility to deliver the automation advantages of cloud services like 
+provisioning, scaling and backup/restore.
+
+### Goals
+Configuring MetalLB layer2 and BGP using kubernetes custom resource "CRD" instead of ConfigMap.
+
+## Proposal
+Using SIG's [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) framework to build new CRD "MetalLB" object 
+and a controller which works as a translator from the user's configuration yaml to a MetalLB's configmap object.
+with this approach there will be no changes required for the MetalLB management plane.
+
+### Risks and Mitigations
+Users need to either use CRD or ConfigMap to provision load balancer instance but not both, if user
+mistakenly used both the CRD configuration will overwrite the user defined ConfigMap.
+
+## Design Details
+![Alt text](./metallb_crd.png?raw=true "MetalLB Kubernetes Controller")
+
+1- User will supply the yaml configuration using the MetalLB resource.
+
+2- Configuration will get authenticated and  authorized based on RBAC rules customized by SIG's framework 
+(it can be further tunned as needed).
+
+3- Supplied configuration will get verified using openAPIV3Schema based on the definition of MetalLB CRD.
+
+4- New objects will be stored in etcd.
+
+5- MetalLB controller monitors API server looking for creation of MetalLB resource.
+
+6- MetalLB controller will use golang based template to convert the configured MetalLB object to ConfigMap object.
+
+7- The generated configmap object will be written to etcd so MetalLB's components can consume it as if the user directly 
+supplied Configmap configuration.
+
+## Testing
+- Manual validation/PoC using KIND cluster
+  - bring up multi node cluster.
+  - deploy metalLB controller and speakers PODs.
+  - deploy MetalLB CRD controller.
+  - apply MetalLB resource config using the following sample configuration.
+    ```bigquery
+    apiVersion: loadbalancer.loadbalancer.operator.io/v1
+    kind: MetalLB
+    metadata:
+      name: metallb-sample
+      namespace: metallb-system
+    spec:
+      address-pools:
+      - name: default
+        protocol: layer2
+        addresses:
+        - 172.18.255.100-172.18.255.200
+    ```
+  - deploy load balancer service application for example NGINX.
+  - verify traffic to the NGINX service.
```

Comment:
Are there any standard Kubernetes e2e tests we could use?

Reply:
not that I am aware of, I am planning to extend existing Metallb e2e test to include MetalLB resource.

Reply:
I dont think there is a MetalLB e2e test as it was dropped recently.